bam_addrprg.c

```diff
@@ -0,0 +1,471 @@
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in
+all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+DEALINGS IN THE SOFTWARE.  */
+
+#include <htslib/sam.h>
+#include "samtools.h"
+#include "sam_opts.h"
+#include <string.h>
+#include <strings.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdbool.h>
+#include <limits.h>
+#include <assert.h>
+#include <unistd.h>
+
+typedef enum {
+    overwrite_all,
+    orphan_only,
+} rg_mode;
+
+struct parsed_opts {
+    char* input_name;
+    char* output_name;
+    char* output_mode;
+    char* rg_id;
+    char* rg_line;
+    rg_mode mode;
+    sam_global_args ga;
+};
+
+struct state;
+typedef struct parsed_opts parsed_opts_t;
+typedef struct state state_t;
+
+struct state {
+    samFile* input_file;
+    bam_hdr_t* input_header;
+    samFile* output_file;
+    bam_hdr_t* output_header;
+    char* rg_id;
+    void (*mode_func)(const state_t*, bam1_t*);
+};
+
+static void cleanup_opts(parsed_opts_t* opts)
+{
+    if (!opts) return;
+    free(opts->rg_id);
+    free(opts->output_mode);
+    free(opts->output_name);
+    free(opts->input_name);
+    sam_global_args_free(&opts->ga);
+    free(opts);
+}
+
+static void cleanup_state(state_t* state)
+{
+    if (!state) return;
+    free(state->rg_id);
+    if (state->output_file) sam_close(state->output_file);
+    bam_hdr_destroy(state->output_header);
+    if (state->input_file) sam_close(state->input_file);
+    bam_hdr_destroy(state->input_header);
+    free(state);
+}
+
+// Converts \t and \n into real tabs and newlines
+static char* basic_unescape(const char* in)
+{
+    assert(in);
+    char *ptr, *out;
+    out = ptr = malloc(strlen(in)+1);
+    size_t size = 0;
+    while (*in) {
+        if (*in == '\\') {
+            ++in;
+            if (*in == '\0') {
+                fprintf(stderr, "Unterminated escape sequence\n");
+                free(out);
+                return NULL;
+            }
+            switch (*in) {
+            case '\\':
+                *ptr = '\\';
+                break;
+            case 't':
+                *ptr = '\t';
+                break;
+            case 'n':
+                *ptr = '\n';
+                break;
```

Comment:
Why are we parsing `\n`?  Supplying multiple lines isn't in the documentation and will lead to confusing things happening, no?